jupyterhub_traefik_proxy/traefik_utils.py

```diff
@@ -0,0 +1,202 @@
+import time
+import socket
+import sys
+import requests
+import etcd3
+from os.path import abspath, dirname, join
+from subprocess import Popen
+from urllib.parse import urlparse
+
+
+_ports = {
+    "traefik": 8000,
+    "default_backend": 9000,
+    "first_backend": 9090,
+    "second_backend": 9099,
+}
+
+
+def get_port(service_name):
+    return _ports[service_name]
+
+
+def is_open(ip, port):
+    timeout = 1
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.settimeout(timeout)
+    try:
+        s.connect((ip, port))
+        s.shutdown(socket.SHUT_RDWR)
+        return True
+    except:
+        return False
+    finally:
+        s.close()
+
+
+def check_host_up(ip, port):
+    """ Allow the service up to 2 sec to open connection on the
+    designated port """
+    up = False
+    retry = 20  # iterations
+    delay = 0.1  # 100 ms
+
+    for i in range(retry):
+        if is_open(ip, port):
+            up = True
+            break
+        else:
+            time.sleep(delay)
+    return up
+
+
+def traefik_routes_to_correct_backend(path, expected_port):
+    """ Check if traefik followed the configuration and routed the
+    request to the right backend """
+    base_url = "http://localhost:" + str(get_port("traefik"))
+    resp = requests.get(base_url + path)
+    assert int(resp.text) == expected_port
+
+
+def check_traefik_etcd_static_conf_ready():
+    base_url = "http://localhost:" + str(get_port("traefik"))
+    """ Allow traefik up to 10 sec to load its static configuration from the
+    etcd cluster """
+    timeout = time.time() + 10
+    ready = False
+    t = 0.1
+    while not ready and time.time() < timeout:
+        resp = requests.get(base_url + "/api/providers/etcdv3")
+        ready = resp.status_code == 200
+        if not ready:
+            t = min(2, t * 2)
+            time.sleep(t)
+    assert ready  # Check that we got here because we are ready
+
+
+def check_traefik_etcd_dynamic_conf_ready(expected_no_of_entries):
+    base_url = "http://localhost:" + str(get_port("traefik"))
```

Comment:
This should be given a url instead of assuming the port here. We cannot assume traefik will always be running on the same port.

This function is going to be useful at runtime because we probably want the `add_route` api to wait for traefik to register the route, not just return immediately when the route has been requested. That means this should probably be a method on the `Proxy` class, e.g.

```python
async def _wait_for_route(routespec):
    ...
```

It's also going to be useful to use `jupyterhub.utils.exponential_backoff` to wait for a condition, which implements some useful things for awaiting conditions:

```python
async def _backend_ready():
    ...
    backends = await get(base_url + ...)
    ...
    return (backend in backends...)
await exponential_backoff(_backend_ready, f"{backend} never arrived")
``